doc/unboxing.md

```diff
@@ -0,0 +1,234 @@
+
+In the future support is expected for:
+
+- Unboxed returns from functions (e.g. returning values of variant type
+  without allocating).
+
+- Removal of allocations from loops where there does not always exist a use
+  of the corresponding unboxed value subsequent to the loop.  This
+  optimisation is expected to be based on partial dead code elimination.
+
+## Control over unboxing
+
+Unboxing depth.
+
+## Decision procedure
+
+The first step is computing the unboxing decision, and then in a second step,
+the decision is used to compute the extra params and args for the
+continuation. This is useful, in order to enable unboxing of recursive
+continuation. Indeed, in the case of a recursive continuation, we will have
+the following scenario: the first step (creating the unboxing decisions) is
+done before entering the handler of the continuation whose parameters we want
+to unbox, and thus we have access to all the non-recursive uses of the
+continuation. Then, once we have finished the downwards pass on the handler
+of the continuation, and thus have finally access to all uses of the
+continuation, we can compute the extra params and args for all use sites.
+
+The creation of unboxing decisions is done using the following steps:
+
+- Using the typing env computed from the continuation uses join, we first
+create an optimistic unboxing decision, based on the shape of the type of each
+of the continuation parameters.
+
+- We then do a pass on each decision which does two things:
+  - pre-compute the extra args for each of the continuation use-site
+  available at that point (i.e. non-recursive uses)
+  - extra_args can be either already in scope, or in some cases need to add
+  an extra primitive (i.e. untag, unbox, block_load, ...)
+
+- If it is not possible to generate such an extra arg (2), remove the
+problematic decision (and associated sub-decisions).
+
+- We then check that for each decision, at least one of the extra args
+computed is "already in scope", else the unboxing is not beneficial.
+
+- Finally, using the decisions, generate the necessary denv to simplify the
+handler.
+
+The second step is simply re-doing the pass the computes the extra args,
+knowing that this time we have all the uses of the continuation, and then
+extending the extra params and args of the continuation.
+
+Another significant change from the current unboxing code is concerning
+variants: whereas before this PR, fields of variants are "shared" across all
+tags (i.e. there is one extra param for the field 0 of all tags, one for all
+fields 1 of all tags, etc...), with this PR, we generate one extra param for
+each pair (field, tag) (i.e. one extra param for the field 0 of tag 0, one
+extra param for field 0 of tag 1, etc...). This required to change row_like
+to consider env_extensions for each possible tag (see the comment at the
+beginning of unbox_continuation_params.ml for more context about why this is
+needed).
```

Comment:
This paragraph probably needs some rewording (talking about "before this PR" is not really a good idea in the context of a permanent doc). I can do it if you want.